### PR TITLE
Update 0.6 to Contracts-v2

### PR DESCRIPTION
**Pull request by ben-chain**
This PR updates all the safety checked (`kall`,`kopy`) functions, and all `*Rewrite` functions, to satisfy the `contracts-v2` interfaces and checks.

### CODE REVIEW

File: libevmasm/JumpdestRemover.cpp [hunk 1]
```diff
@@ -56,6 +56,7 @@ bool JumpdestRemover::optimise(set<size_t> const& _tagsReferencedFromOutside)
 set<size_t> JumpdestRemover::referencedTags(AssemblyItems const& _items, size_t _subId)
 {
 	set<size_t> ret;
+
```
Comment (K-Ho) [inline libevmasm/JumpdestRemover.cpp]: [nit] remove

File: CMakeLists.txt [hunk 2]
```diff
@@ -12,7 +12,7 @@ eth_policy()
 # project name and version should be set after cmake_policy CMP0048
 set(PROJECT_VERSION "0.6.12")
 # OSX target needed in order to support std::visit
-set(CMAKE_OSX_DEPLOYMENT_TARGET "10.14")
+set(CMAKE_OSX_DEPLOYMENT_TARGET "10.15")
```
Comment (K-Ho) [inline CMakeLists.txt:15]: I just built fine with this at 10.14, you sure we need this?

File: libevmasm/PeepholeOptimiser.cpp [hunk 3]
```diff
@@ -126,7 +126,12 @@ struct OpPop: SimplePeepholeOptimizerMethod<OpPop, 2>
 		std::back_insert_iterator<AssemblyItems> _out
 	)
 	{
-		if (_pop == Instruction::POP && _op.type() == Operation)
+		
+		if (
+			_pop == Instruction::POP && _op.type() == Operation
+			// OVM: do not optimize CALLER POP as this is how we identify safe call to 0x04 precompile. ("kopy" in EVMDialect)
+			&& _op.instruction() != Instruction::CALLER
```
Comment (K-Ho) [inline libevmasm/PeepholeOptimiser.cpp:133]: 📞 🍾( CALLER POP)

File: libsolidity/codegen/CompilerContext.cpp [hunk 4]
```diff
@@ -138,18 +136,41 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
 	m_disable_rewrite = true;
 
 	auto callYUL = R"(
-		mstore(add(callBytes, 4), addr)
+		// store _gasLimit
+		mstore(add(callBytes, 0x04), in_gas)
+		// store _address
+		mstore(add(callBytes, 0x24), addr)
+		// store abi byte memory offset bullshit
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: watch yo profamity

File: libsolidity/codegen/CompilerContext.cpp [hunk 5]
```diff
@@ -138,18 +136,41 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
...
-			mstore(add(add(callBytes, 0x24), ptr), mload(add(argsOffset, ptr)))
+			mstore(add(rawCallBytes, ptr), mload(add(argsOffset, ptr)))
 		}
-		let success := kall(callBytes, add(0x24, argsLength), retOffset, retLength)
-		if eq(success, 0) {
-			let result := mload(0x40)
-			returndatacopy(result, 0, returndatasize())
-			revert(result, returndatasize())
+		// we will return to where the calldata ends.  Overpadded to ensure that we do sufficient trailing zeros for ABI encoding
+		let endOfArgs := add(callBytes, add(0x104, argsLength))
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: can we change this 0x104 -> 0x20? This is just to account for argsLength not ending directly at the end of a word right?

File: libsolidity/codegen/CompilerContext.cpp [hunk 6]
```diff
@@ -165,17 +186,44 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
...
+				);
+				ret = false;
+				break;
+			case Instruction::BLOCKHASH:
+			case Instruction::CALLCODE:
+			case Instruction::COINBASE:
+			case Instruction::DIFFICULTY:
+			case Instruction::GASPRICE:
+			case Instruction::ORIGIN:
+			case Instruction::SELFBALANCE:
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: Move `SELFBALANCE` to above `BALANCE` so it has the same error message

File: libsolidity/codegen/CompilerContext.cpp [hunk 7]
```diff
@@ -188,57 +236,103 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
...
-						if eq(success, 0) { revert(0, 0) }
+						// technically 0x64 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x84, length), callBytes, 0x20)
+						// salt is first stack val in ==> first stack val out (address)
 						salt := mload(callBytes)
+
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: Shouldn't this just be `mstore(ptr, 0x00)`

File: libsolidity/codegen/CompilerContext.cpp [hunk 8]
```diff
@@ -188,57 +236,103 @@ bool CompilerContext::appendCallback(evmasm::AssemblyItem const& _i) {
...
-						if eq(success, 0) { revert(0, 0) }
+						// technically 0x44 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x64, length), callBytes, 0x20)
+						// legnth is first stack val in ==> first stack val out (address)
 						length := mload(callBytes)
+
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: Should this be `mstore(ptr, 0x00)` ?

**Review by K-Ho — APPROVED** [review]
lgtm! couple notes

File: CMakeLists.txt (hunk 2 above)
Comment (ben-chain) [reply CMakeLists.txt:15]: I've had issues with this when jumping between 0.5 and 0.6 branches; I'm gonna leave unless you think it breaks when you try it

File: libsolidity/codegen/CompilerContext.cpp (hunk 7 above)
Comment (ben-chain) [reply libsolidity/codegen/CompilerContext.cpp]: good catch

File: libsolidity/codegen/CompilerContext.cpp (hunk 8 above)
Comment (ben-chain) [reply libsolidity/codegen/CompilerContext.cpp]: good catch

**Pull request merged**